### DISCUSSION TITLE
Arduino based MIDI Bank and Voice Selector

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#236] [Show and tell]
I've put together the code for an Arduino based bank and voice selector using some really cheap 8-way button modules I found online.  The idea is to have something that would mirror the kind of button configuration you'd have on a real DX synth, although for now it is just a mess of spaghetti wiring on a breadboard!

I've tested it with an Uno using serial MIDI but it should also work on a Pro Micro (or similar) with USB MIDI if required.

If that sounds of interest, you read more about it here: https://diyelectromusic.wordpress.com/2022/05/17/arduino-midi-button-voice-select/

Kevin

![ArduinoMIDIButtonVoiceSelect](https://user-images.githubusercontent.com/68612569/168913950-bcddfb0a-2acb-440f-a09f-6773ff7b4549.png)

**Comment by probonopd** [comment]
Interesting. Maybe you can put the code onto the same Raspberry Pi Pico that can also be used at the same time as a MIDI-to-Serial bridge for the PC... saving one component for those who need that functionality.

**Reply by izash** [reply]
Adding the code to the Pi Pico is a great idea.

**Reply by diyelectromusic** [reply]
Well I have a project that is part way there for that, but it would need code to handle the "matrix" side of things: https://diyelectromusic.wordpress.com/2022/05/04/pi-pico-midi-button-controller/

My take on a Pico USB MIDI to serial MIDI is here: https://diyelectromusic.wordpress.com/2022/04/21/circuitpython-usb-to-serial-midi-router/ so I expect some combination would be possible!

Alternatively, presumably the Pico is acting as a USB device to the MiniDexed, outputting serial MIDI to a PC?  If so, then adding simple software THRU to a Pro Micro would do the same?  I guess it depends on which direction things are going.  All my Arduino MIDI USB projects are listed here, so there is probably the right "bit" in there somewhere: https://diyelectromusic.wordpress.com/2022/03/22/arduino-and-usb-midi/

Kevin

**Reply by probonopd** [reply]
> presumably the Pico is acting as a USB device to the MiniDexed, outputting serial MIDI to a PC? 

No, the other way around: The Pico is acting as a USB MIDI device to the PC and connects to 3.3V TX, RX and GND on the MiniDexed device. MiniDexed is the "3V3 MIDI" in yoiur picture:

![image](https://user-images.githubusercontent.com/2480569/169204469-458c0992-4f00-4c5f-a415-2afbda6d89e3.png)

By the way, the 3V3 line must NOT be connected to the MiniDexed device since MiniDexed has its own power and the RPI Pico gets powered over USB. Just mentioning here in case others are reading this.

Using MicroPython for this is smart, because it allows for easy customization to make use of all those nice extra pins on the Pico.

> The only slight quirk is that I try to filter out unrecognised MIDI messages to limit the forwarded traffic

Is this really absolutely needed? I would try to do without this in order not to introduce addtional error sources for sysex, making it a "transparent bridge". I'd assume that the Zero is plenty powerful enough.

**Reply by diyelectromusic** [reply]
> Is this really absolutely needed? I would try to do without this in order not to introduce addtional error sources for sysex, making it a "transparent bridge". I'd assume that the Zero is plenty powerful enough.

Unfortunately as I understand things (my python isn't great) this is a consequence of how the Adafruit MIDI library works.  Each MIDI message is an object in its own right, and anything that isn't imported into your code comes up as "MIDIUnknownEvent".  But if you try to send a "MIDIUnknownEvent" you get a run-time TypeError in midi_message.py...  This is why all the different types of MIDI messages have to be imported at the start.

To be honest, I really, really don't like how the Adafruit MIDI library works... but for most simple applications it is quite easy for "beginners" hacking away in Python I guess...

Maybe someone who knows python could have a look and let me know if there is a better way around this or if I'm missing something?

I might try to go back to my own MIDI decoder (I wrote a simple one for Micropython), but I'm not sure how the USB MIDI would work without the rest of the Adafruit MIDI library - they seem quite well linked together.

**Reply by diyelectromusic** [reply]
Ok, so given the proviso about about the Adafruit MIDI library and unrecognised events, here is the Pico version with a video of me playing one my Lo-Fi Orchestra MIDI files through the Pico to the MiniDexed from a PC over USB whilst changing voices and banks.

https://diyelectromusic.wordpress.com/2022/05/20/pi-pico-midi-button-voice-select/

If there is someone more pythonic than me who wants to take a look and tell me how I ought to be handling the Adafruit MIDI objects, then please do chip in! :)

Kevin

**Reply by probonopd** [reply]
This is begging for a 3d printed case! ;-)

**Reply by diyelectromusic** [reply]
> This is begging for a 3d printed case! ;-)

I want someone to give me a button bank like you get on a DX11 :)

**Comment by DutchBigEd** [comment]
Hey, this is nice, I’m working on something similar, ordered 2x (4x4)switches and 2x (2x4) switches @ probably the same address as you did 🤭
I will be using an Arduino Uno for reading the buttons as well as 8 potmeters (for controlling the volume for each instance).
The Uno is programmed using Hiduino (google) so it is recognized as a native midi interface.

**Reply by DutchBigEd** [reply]
![20230121_170028](https://user-images.githubusercontent.com/17274454/213875601-f077f203-ef0f-4f0b-8809-0854b788ef18.jpg)
My build will be housed in a 1/2 19" box

**Reply by diyelectromusic** [reply]
Looks amazing :)  I'm working on a "TX816" style MiniDexed too!  Watch this space!

Kevin

**Reply by DutchBigEd** [reply]
I will!

**Reply by diyelectromusic** [reply]
> Looks amazing :) I'm working on a "TX816" style MiniDexed too! Watch this space!

Ok, I've written up and documented all the "theory" so far:
Part 1 provides the general background, context and aims.
Part 2 looks at the PCB designs I'm going to try to use.
Part 3 looks at a front panel design that tries to keep the TX816 esthetic.

If you're interested, see: https://diyelectromusic.wordpress.com/2023/01/22/minidexed-tx816/

Now I have all the bits back, I need to start to put it together and see what I've forgotten.  The chances of everything going together first time and actually doing something useful are pretty slim, but stranger things have been known to happen!  But that will have to wait now until the next "project time slot" (as it were).

So back to "watch this space" again for now :)

Kevin

![IMG_6819](https://user-images.githubusercontent.com/68612569/213944612-bcec24b9-72d3-406e-b14e-d58716932175.JPG)

**Reply by probonopd** [reply]
Wow, you are _really_ determined. I am speechless! :1st_place_medal:

**Reply by diyelectromusic** [reply]
For what its worth, I have the PCB build guides up now, and think I've found all the mistakes I've made!

I'm starting to integrate the boards now and have managed to get the Pico to drive all the IO.  Next will be starting to think about the full MiniDexed side of things... and then I'll hit all the performance problems!

But I'm getting there slowly :)

Kevin

https://youtu.be/0zjuh_fSTmI

**Reply by diyelectromusic** [reply]
I'm starting to get somewhere now I think. I'm not entirely convinced that Micropython gives me the performance I need, but it isn't too bad... at least it is routing MIDI now!

Still plenty to do mind...

Here is some lo-fi 80s DX7 FM synth "Jupiter" from Holst's Planets suite...  (and yes, you can just do this with MiniDexed out of the box of course, but I can select voices on the fly here, and I quite like the aesthetics :))

https://youtu.be/n0GQlM5xTSY

Kevin
